### PR TITLE
test/network: echo integration test ipv6

### PR DESCRIPTION
**Pull request by hennna**
This PR parameterizes an echo server integration test and is a redo of #883 . The differences between this PR and #883 are as follows:

- Dual bind turn off was split off into a separate PR

- Add admin ipv6 support

- Add fake upstream ipv6 support

### CODE REVIEW

File: source/common/network/utility.cc [hunk 1]
```diff
@@ -183,7 +198,14 @@ bool Utility::isLoopbackAddress(const Address::Instance& address) {
     return false;
   }
 
-  return address.ip()->ipv4()->address() == htonl(INADDR_LOOPBACK);
+  if (address.ip()->version() == Address::IpVersion::v4) {
+    // Compare to the canonical v4 loopback address: 127.0.0.1.
+    return address.ip()->ipv4()->address() == htonl(INADDR_LOOPBACK);
+  } else if (address.ip()->version() == Address::IpVersion::v6) {
+    std::array<uint8_t, 16> addr = address.ip()->ipv6()->address();
+    return 0 == memcmp(&addr, &in6addr_loopback, sizeof(struct in6_addr));
```
Comment (htuch) [inline source/common/network/utility.cc]: `sizeof(addr)` or `sizeof(in6addr_loopback)`.

File: source/common/network/utility.cc [hunk 2]
```diff
@@ -157,6 +164,14 @@ Address::InstanceConstSharedPtr Utility::getLocalAddress() {
     freeifaddrs(ifaddr);
   }
 
+  // If the local address is not found above, then return the loopback addresss by default.
```
Comment (htuch) [inline source/common/network/utility.cc]: Can you add a reference to https://github.com/lyft/envoy/issues/939?

File: source/common/network/utility.cc [hunk 3]
```diff
@@ -183,7 +198,14 @@ bool Utility::isLoopbackAddress(const Address::Instance& address) {
...
 
-  return address.ip()->ipv4()->address() == htonl(INADDR_LOOPBACK);
+  if (address.ip()->version() == Address::IpVersion::v4) {
+    // Compare to the canonical v4 loopback address: 127.0.0.1.
+    return address.ip()->ipv4()->address() == htonl(INADDR_LOOPBACK);
+  } else if (address.ip()->version() == Address::IpVersion::v6) {
+    std::array<uint8_t, 16> addr = address.ip()->ipv6()->address();
+    return 0 == memcmp(&addr, &in6addr_loopback, sizeof(struct in6_addr));
+  }
+  return false;
```
Comment (htuch) [inline source/common/network/utility.cc]: `NOT_IMPLEMENTED;`

File: test/integration/integration_admin_test.cc [hunk 4]
```diff
@@ -4,6 +4,7 @@
 
 #include "test/integration/integration_test.h"
 #include "test/integration/utility.h"
+#include "test/test_common/environment.h"
```
Comment (htuch) [inline test/integration/integration_admin_test.cc]: Why is this needed (no other changes to this file)?

File: test/integration/integration_test.cc [hunk 5]
```diff
@@ -5,6 +5,7 @@
 #include "common/http/header_map_impl.h"
 
 #include "test/integration/utility.h"
+#include "test/test_common/environment.h"
```
Comment (htuch) [inline test/integration/integration_test.cc]: Ditto.

File: source/common/network/utility.cc [hunk 6]
```diff
@@ -129,7 +129,11 @@ uint32_t Utility::portFromTcpUrl(const std::string& url) {
   }
 }
 
-Address::InstanceConstSharedPtr Utility::getLocalAddress() {
+// TODO(hennna): Currently getLocalAddress does not support choosing between
+// multiple interfaces and addresses not returned by getifaddrs. In additon,
+// the default is to return a loopback address of type version. This fucntion may
```
Comment (mattklein123) [inline source/common/network/utility.cc]: typo fucntion

File: source/common/network/utility.cc [hunk 7]
```diff
@@ -129,7 +129,11 @@ uint32_t Utility::portFromTcpUrl(const std::string& url) {
...
+// need to be updated in the future.
+Address::InstanceConstSharedPtr Utility::getLocalAddress(const Address::IpVersion version) {
```
Comment (mattklein123) [inline source/common/network/utility.cc]: IIRC server.cc (or somewhere) has error checking that this function returns a null address, which can't happen anymore. Can you fix the callers to remove error checking that can't happen?

File: source/exe/main.cc [hunk 8]
```diff
@@ -59,8 +59,10 @@ int main(int argc, char** argv) {
   DefaultTestHooks default_test_hooks;
   Stats::ThreadLocalStoreImpl stats_store(*restarter);
   Server::ProdComponentFactory component_factory;
-  LocalInfo::LocalInfoImpl local_info(Network::Utility::getLocalAddress(), options.serviceZone(),
-                                      options.serviceClusterName(), options.serviceNodeName());
+  // TODO(henna): Parameterize local address IP version.
```
Comment (mattklein123) [inline source/exe/main.cc]: nit: less parameterize, more probably add CLI option.

File: test/config/integration/echo_server.json [hunk 9]
```diff
@@ -0,0 +1,23 @@
+{
```
Comment (mattklein123) [inline test/config/integration/echo_server.json]: Since you are moving this test into a standalone config, is there any reason to leave the echo listener in the other config? Seems like that should be removed?

File: test/config/integration/echo_server.json [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+          "domain": "foo",
+          "descriptors": [[{"key": "foo", "value": "bar"}]],
+          "stat_prefix": "name"
+        }
+      },
+      { "type": "read", "name": "echo", "config": {} }
+    ]
+  }],
+
+  "admin": { "access_log_path": "/dev/null", "profile_path": "{{ test_tmpdir }}/envoy.prof", "address": "tcp://{{ ip_loopback_address }}:0" },
```
Comment (mattklein123) [inline test/config/integration/echo_server.json]: nit: can we break this at 100 cols

File: test/integration/utility.h [hunk 11]
```diff
@@ -22,7 +23,10 @@ class BufferingStreamDecoder : public Http::StreamDecoder, public Http::StreamCa
   BufferingStreamDecoder(std::function<void()> on_complete_cb) : on_complete_cb_(on_complete_cb) {}
 
   bool complete() { return complete_; }
-  const Http::HeaderMap& headers() { return *headers_; }
+  const Http::HeaderMap& headers() {
+    ASSERT(headers_ != nullptr);
```
Comment (mattklein123) [inline test/integration/utility.h]: nit: per style guide generally shy away from this type of ASSERT check because it should crash in an obvious way very soon.

Comment (hennna) [inline test/integration/utility.h]: STR_EQ comparisons in the integration tests like this one : https://github.com/lyft/envoy/blob/master/test/integration/server.cc#L60 cause a segfault if response->complete() is false. Would it make more sense to ASSERT_TRUE(response->complete()) and remove this change? The only problem is that this particular example is in the destructor of IntegrationTestServer.

Comment (mattklein123) [inline test/integration/utility.h]: IMO a segfaulting test is not the end of the world as it failed. :) I would probably just do nothing, but if you want to do something more clear I would use a gtest assert rather than the prod assert. So ASSET_NE or something.

Comment (htuch) [inline test/integration/utility.h]: Arguably, since we're returning a reference here, if it was nullptr the failure would be harder to understand the further away from this return we get, so best to check here.

Comment (mattklein123) [inline test/integration/utility.h]: Better would probably be to just make this return a pointer instead of a reference, but that's a larger change. I don't feel very strongly about this either way.

Comment (hennna) [inline test/integration/utility.h]: Ok. I'll just leave as is for now (no change) and leave the change for another PR.
